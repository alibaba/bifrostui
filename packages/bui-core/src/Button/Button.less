@import '~@bifrostui/styles/mixins/index.less';

.btn-icon() {
  display: inline-flex;
  align-self: center;
  flex-shrink: 0;
}

.btn-outline(@color) {
  color: var(~'--bui-color-@{color}');
  border-color: var(~'--bui-color-@{color}');
}

.btn-contained(@color) {
  color: var(--bui-color-white);
  .linear-gradient(var(~'--bui-color-@{color}-start'), var(~'--bui-color-@{color}-end'));
}

.btn-light(@color) {
  background-image: none;
  border: 1px solid transparent;
  color: var(~'--bui-color-@{color}');
  background-color: var(~'--bui-color-@{color}-light');
}

.btn-text(@color) {
  color: var(~'--bui-color-@{color}');
}

.bui-btn {
  --bg-color: var(--bui-button-default-bg-color, var(--bui-color-neutral-5));
  --text-color: var(--bui-button-text-color, var(--bui-color-fg-muted));
  --border-color: var(--bui-button-border-color, var(--bui-color-neutral-4));
  --border-radius: var(--bui-button-border-radius, 100px);
  --font-weight: var(--bui-button-font-weight, var(--bui-font-weight-medium));
  border: 1px solid transparent;
  outline: 0;
  background-color: transparent;
  cursor: pointer;
  color: var(--text-color);
  box-sizing: border-box;
  position: relative;
  display: inline-flex;
  margin: 0;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
  font-size: var(--bui-text-size-3);
  font-weight: var(--font-weight);
  border-radius: var(--border-radius);
  height: 27px;
  line-height: var(--bui-line-height);
  font-family: var(--bui-font-family);

  @keyframes spin {
    from {
      transform: rotate(0deg);
    }

    to {
      transform: rotate(360deg);
    }
  }

  &::after {
    border: 0;
  }

  &::before {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 100%;
    height: 100%;
    background-color: var(--bui-color-black);
    border: inherit;
    border-color: var(--bui-color-black);
    border-radius: inherit;
    transform: translate(-50%, -50%);
    opacity: 0;
    content: ' ';
  }

  &:active::before {
    opacity: 0.04;
  }

  &-icon-start {
    .btn-icon();
    margin-right: var(--bui-spacing-xs);
  }

  &-icon-end {
    .btn-icon();
    margin-left: var(--bui-spacing-xs);
  }

  &-disabled {
    pointer-events: none;
    opacity: 0.5;
  }

  &-small {
    font-size: var(--bui-text-size-4);
    padding: 0 11px;
    height: 24px;
  }

  &-medium {
    padding: 0 14px;
    height: 27px;
  }

  &-large {
    font-size: var(--bui-text-size-2);
    padding: 0 14px;
    height: 33px;
  }

  &-full {
    font-size: var(--bui-title-size-4);
    width: 100%;
    height: 42px;
  }

  &-text {
    &.bui-btn-primary {
      .btn-text(primary);
    }

    &.bui-btn-success {
      .btn-text(success);
    }

    &.bui-btn-info {
      .btn-text(info);
    }

    &.bui-btn-warning {
      .btn-text(warning);
    }

    &.bui-btn-danger {
      .btn-text(danger);
    }

    &.bui-btn-vip {
      .btn-text(vip);
    }
  }

  &-contained {
    background-color: var(--bg-color);

    &.bui-btn-primary {
      .btn-contained(primary);
    }

    &.bui-btn-success {
      .btn-contained(success);
    }

    &.bui-btn-info {
      .btn-contained(info);
    }

    &.bui-btn-warning {
      .btn-contained(warning);
    }

    &.bui-btn-danger {
      .btn-contained(danger);
    }

    &.bui-btn-vip {
      color: #582331;
      background-image: linear-gradient(
        90deg,
        var(--bui-color-vip-start) 3%,
        var(--bui-color-vip-end) 100%
      );
    }
  }

  &-outlined {
    border: solid 1px var(--border-color);

    &.bui-btn-primary {
      .btn-outline(primary);
    }

    &.bui-btn-success {
      .btn-outline(success);
    }

    &.bui-btn-info {
      .btn-outline(info);
    }

    &.bui-btn-warning {
      .btn-outline(warning);
    }

    &.bui-btn-danger {
      .btn-outline(danger);
    }

    &.bui-btn-vip {
      .btn-outline(vip);
    }
  }

  &-light {
    background-image: none;
    background-color: var(--bui-color-gray-light);
    border: 1px solid transparent;

    &.bui-btn-warning {
      .btn-light(warning);
    }

    &.bui-btn-primary {
      .btn-light(primary);
    }

    &.bui-btn-info {
      .btn-light(info);
    }

    &.bui-btn-success {
      .btn-light(success);
    }

    &.bui-btn-danger {
      .btn-light(danger);
    }

    &.bui-btn-vip {
      color: var(--bui-color-vip);
      position: relative;
      background-color: fade(#ff866e, 30%);
    }
  }
}
