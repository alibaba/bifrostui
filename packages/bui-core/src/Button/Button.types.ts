import { OverrideProps, ThemeColor, ButtonOpenType } from '@bifrostui/types';
import React from 'react';

export type ButtonProps<
  D extends React.ElementType = 'button',
  P = {},
> = OverrideProps<
  {
    props: P & {
      /**
       * 是否禁用
       * @default false
       */
      disabled?: boolean;
      /**
       * 按钮尺寸
       * - small: 小按钮
       * - medium: 中号按钮，默认按钮
       * - large: 大按钮/半屏按钮
       * - full: 通栏按钮
       * @default 'medium'
       */
      size?: 'small' | 'medium' | 'large' | 'full';
      /** 按钮颜色
       * @default 'default'
       */
      color?: ThemeColor | 'default' | 'vip';
      /**
       * 按钮类型
       * - text: 文字按钮
       * - outlined: 描边按钮
       * - contained: 实心填充按钮
       * - light: 浅亮填充按钮
       * @default 'outlined'
       */
      variant?: 'outlined' | 'text' | 'contained' | 'light';
      /**
       * 在子元素前放在的内容
       */
      startIcon?: React.ReactNode;
      /**
       * 在子元素后放在的内容
       */
      endIcon?: React.ReactNode;
      /**
       * 点击事件回调方法
       * @param event
       * @returns
       */
      onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
      /**
       * 仅微信小程序支持的button开放能力
       */
      openType?: ButtonOpenType;
    };
    defaultComponent: D;
  },
  D
>;
