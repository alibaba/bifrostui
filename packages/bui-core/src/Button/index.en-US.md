---
group:
  title: Basic input
  order: -1
order: -1
nav: assembly
name: Button button
---

# Button button

The button is used to trigger an operation, make a selection, such as submitting a form, opening a dialog box, canceling the operation, or performing a deletion operation.

## Code demonstration

### Basic button

The button supports four display forms: text button, stroke button, fill button, and light button. The default is the stroke button.

Text button: Use text directly as a button. It is the button with the weakest visual appeal, usually appearing next to table operation bars, titles, and fields.

Fill button: generally used for the main button, it is the highest priority operation that users need to pay attention to on the entire page, guiding users to pay attention to and operate.

Stroke button: a common form of white background with text, which is visually less emphasized than the fill button and is usually used in combination with the fill button

Light button: The text in light state is the background color of the button, and the background is the corresponding light color.

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="8px" flexWrap="wrap">
      <Button color="primary" variant="text">
        文字按钮
      </Button>
      <Button color="primary">描边按钮</Button>
      <Button color="primary" variant="contained">
        填充按钮
      </Button>
      <Button color="primary" variant="subtle">
        浅亮按钮
      </Button>
    </Stack>
  );
};
```

### Button size

The buttons come in three sizes: large, medium, and small.

Set the 'size' to 'large small full' and set the button size to 'large', 'small', and 'full' respectively. If 'size' is not set, the size will be medium.

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack spacing="8px">
      <Stack direction="row" spacing="8px" style={{ width: '350px' }}>
        <Button size="small">小号按钮</Button>
        <Button size="medium">中号按钮</Button>
        <Button size="large">大号按钮</Button>
      </Stack>
      <Button size="full" style={{ width: '350px' }}>
        通栏按钮
      </Button>
      <Stack direction="row" spacing="8px" style={{ width: '350px' }}>
        <Button size="small" variant="contained" color="primary">
          小号按钮
        </Button>
        <Button size="medium" variant="contained" color="primary">
          中号按钮
        </Button>
        <Button size="large" variant="contained" color="primary">
          大号按钮
        </Button>
      </Stack>
      <Button
        size="full"
        variant="contained"
        color="primary"
        style={{ width: '350px' }}
      >
        通栏按钮
      </Button>
    </Stack>
  );
};
```

### Button color

The button supports' primary ' Definition of themes such as success, information, warning, danger, and VIP. If color is not set, it defaults to 'default'

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="8px" flexWrap="wrap">
      <Button variant="contained">取消</Button>
      <Button color="primary" variant="contained">
        购票
      </Button>
      <Button color="success" variant="contained">
        影城
      </Button>
      <Button color="info" variant="contained">
        电影
      </Button>
      <Button color="warning" variant="contained">
        想看
      </Button>
      <Button color="danger" variant="contained">
        特惠
      </Button>
      <Button color="vip" variant="contained">
        会员
      </Button>
    </Stack>
  );
};
```

### Disabled status

Disable the button through the 'disabled' attribute, and the button cannot be clicked in the disabled state.

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="8px">
      <Button variant="text" disabled>
        文字禁用
      </Button>
      <Button disabled>描边禁用</Button>
      <Button variant="contained" disabled>
        填充禁用
      </Button>
    </Stack>
  );
};
```

### Button with icon

Because users are more sensitive to icons compared to plain text, sometimes you may want to set icons for certain buttons to enhance the user experience of the application. For example, if you have a date button, you can use a date icon to mark it StartIcon 'is located in front of the child element EndIcon 'is located after the child element.

```tsx
import React from 'react';
import { Button, Stack } from '@bifrostui/react';
import { CalendarOutlinedIcon, HeartFilledIcon } from '@bifrostui/icons';

export default () => {
  return (
    <Stack direction="row" spacing="8px">
      <Button
        startIcon={
          <HeartFilledIcon htmlColor="linear-gradient(150deg, #FFFDFC 0%, #D5B198 100%)" />
        }
      >
        Heart
      </Button>
      <Button endIcon={<CalendarOutlinedIcon />}>Date</Button>
    </Stack>
  );
};
```

## API

| attribute | explain                                     | type                                                                      | Default value |
| --------- | ------------------------------------------- | ------------------------------------------------------------------------- | ------------- |
| variant   | Button type                                 | `outlined` \|`text` \|`contained` \|`light`                               | `outlined`    |
| color     | Button color                                | `primary` \|`info` \|`success` \|`warning` \|`danger` \|`vip` \|`default` | `default`     |
| size      | Button size                                 | `small` \|`medium` \|`large` \|`full`                                     | `medium`      |
| disabled  | Do you want to disable it                   | boolean                                                                   | false         |
| startIcon | The content placed before the child element | React.ReactNode                                                           | -             |
| endIcon   | The content placed after the child element  | React.ReactNode                                                           | -             |
| openType  | WeChat Open Capability                      | string                                                                    | -             |
| onClick   | Callback when clicking a button             | (event: MouseEvent) => void                                               | -             |

For other properties see [button](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attributes)
