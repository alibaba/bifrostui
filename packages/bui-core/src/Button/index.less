@import '@bifrostui/styles/mixins/index.less';

@{root-selector} {
  --bui-btn-font-size: var(--bui-text-size-3);
  --bui-btn-font-weight: var(--bui-font-weight-medium);
  --bui-btn-border-radius: 120px;
  --bui-btn-small-border-radius: 120px;
  --bui-btn-large-border-radius: 120px;
  --bui-btn-full-border-radius: 120px;
  --bui-btn-color: var(--bui-color-fg-muted);
  --bui-btn-bg: var(--bui-color-bg-view);
  --bui-btn-border: 0;
  --bui-btn-height: 27px;
  --bui-btn-small-height: 24px;
  --bui-btn-large-height: 33px;
  --bui-btn-full-height: 42px;
  --bui-btn-subtle-border: 0;
  --bui-btn-outlined-border: 1px solid var(--bui-color-neutral-4);
  --bui-btn-disabled-opacity: 0.5;
  --bui-btn-small-padding: 0 11px;
  --bui-btn-padding: 0 14px;
  --bui-btn-large-padding: 0 14px;
  --bui-btn-full-font-size: var(--bui-title-size-4);
  --bui-btn-full-font-family: var(--bui-font-family);
}

.btn-icon() {
  display: inline-flex;
  align-self: center;
  flex-shrink: 0;
}

.btn-outline(@color) {
  color: var(~'--bui-color-@{color}');
  border-color: var(~'--bui-color-@{color}');
}

.btn-contained(@color) {
  color: var(--bui-color-white);
  .linear-gradient(var(~'--bui-color-@{color}-start'), var(~'--bui-color-@{color}-end'));
}

.btn-subtle(@color) {
  color: var(~'--bui-color-@{color}');
  background: var(~'--bui-color-@{color}-light');
}

.btn-text(@color) {
  color: var(~'--bui-color-@{color}');
}

.bui-btn {
  outline: 0;
  cursor: pointer;
  box-sizing: border-box;
  position: relative;
  display: inline-flex;
  margin: 0;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
  color: var(--bui-btn-color);
  background: var(--bui-btn-bg);
  font-size: var(--bui-btn-font-size);
  font-weight: var(--bui-btn-font-weight);
  border: var(--bui-btn-border);
  border-radius: var(--bui-btn-border-radius);
  height: var(--bui-btn-height);
  padding: var(--bui-btn-padding);
  line-height: var(--bui-line-height);
  font-family: var(--bui-font-family);

  @keyframes spin {
    from {
      transform: rotate(0deg);
    }

    to {
      transform: rotate(360deg);
    }
  }

  &::after {
    border: 0;
  }

  &::before {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 100%;
    height: 100%;
    background-color: var(--bui-color-black);
    border: inherit;
    border-color: var(--bui-color-black);
    border-radius: inherit;
    transform: translate(-50%, -50%);
    opacity: 0;
    content: ' ';
  }

  &:active::before {
    opacity: 0.04;
  }

  &-icon-start {
    .btn-icon();
    margin-right: var(--bui-spacing-xs);
  }

  &-icon-end {
    .btn-icon();
    margin-left: var(--bui-spacing-xs);
  }

  &-disabled {
    pointer-events: none;
    opacity: var(--bui-btn-disabled-opacity);
  }

  &-small {
    font-size: var(--bui-text-size-4);
    padding: var(--bui-btn-small-padding);
    height: var(--bui-btn-small-height);
    border-radius: var(--bui-btn-small-border-radius);
  }

  &-large {
    font-size: var(--bui-text-size-2);
    padding: var(--bui-btn-large-padding);
    height: var(--bui-btn-large-height);
    border-radius: var(--bui-btn-large-border-radius);
  }

  &-full {
    font-family: var(--bui-btn-full-font-family);
    font-size: var(--bui-btn-full-font-size);
    width: 100%;
    height: var(--bui-btn-full-height);
    border-radius: var(--bui-btn-full-border-radius);
  }

  &-text {
    background: none;

    &.bui-btn-primary {
      .btn-text(primary);
    }

    &.bui-btn-success {
      .btn-text(success);
    }

    &.bui-btn-info {
      .btn-text(info);
    }

    &.bui-btn-warning {
      .btn-text(warning);
    }

    &.bui-btn-danger {
      .btn-text(danger);
    }

    &.bui-btn-light {
      color: var(--bui-color-bg-view);
    }

    &.bui-btn-dark {
      color: var(--bui-color-fg-default);
    }

    &.bui-btn-vip {
      .btn-text(vip);
    }
  }

  &-contained {
    background: var(--bui-color-neutral-5);

    &.bui-btn-primary {
      .btn-contained(primary);
    }

    &.bui-btn-success {
      .btn-contained(success);
    }

    &.bui-btn-info {
      .btn-contained(info);
    }

    &.bui-btn-warning {
      .btn-contained(warning);
    }

    &.bui-btn-danger {
      .btn-contained(danger);
    }

    &.bui-btn-light {
      color: var(--bui-color-fg-default);
      background: var(--bui-color-bg-view);
    }

    &.bui-btn-dark {
      color: var(--bui-color-bg-view);
      background: var(--bui-color-fg-default);
    }

    &.bui-btn-vip {
      color: #582331;
      background-image: linear-gradient(
        90deg,
        var(--bui-color-vip-start) 3%,
        var(--bui-color-vip-end) 100%
      );
    }
  }

  &-outlined {
    background: none;
    border: var(--bui-btn-outlined-border);

    &.bui-btn-primary {
      .btn-outline(primary);
    }

    &.bui-btn-success {
      .btn-outline(success);
    }

    &.bui-btn-info {
      .btn-outline(info);
    }

    &.bui-btn-warning {
      .btn-outline(warning);
    }

    &.bui-btn-danger {
      .btn-outline(danger);
    }

    &.bui-btn-light {
      color: var(--bui-color-bg-view);
      border: 1px solid var(--bui-color-bg-view);
    }

    &.bui-btn-dark {
      color: var(--bui-color-fg-default);
      border: 1px solid var(--bui-color-fg-default);
    }

    &.bui-btn-vip {
      .btn-outline(vip);
    }
  }

  &-subtle {
    background-image: none;
    background-color: var(--bui-color-gray-light);
    border: var(--bui-btn-subtle-border);

    &.bui-btn-warning {
      .btn-subtle(warning);
    }

    &.bui-btn-primary {
      .btn-subtle(primary);
    }

    &.bui-btn-info {
      .btn-subtle(info);
    }

    &.bui-btn-success {
      .btn-subtle(success);
    }

    &.bui-btn-danger {
      .btn-subtle(danger);
    }

    &.bui-btn-light {
      color: var(--bui-color-bg-view);
    }

    &.bui-btn-dark {
      color: var(--bui-color-fg-default);
    }

    &.bui-btn-vip {
      color: var(--bui-color-vip);
      position: relative;
      background-color: fade(#ff866e, 30%);
    }
  }
}
