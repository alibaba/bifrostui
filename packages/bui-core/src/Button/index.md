---
group:
  title: 基本输入
  order: -1
order: -1
nav: 基础组件
name: Button 按钮
---

# Button 按钮

按钮用于触发一个操作，做出选择，例如提交表单，打开对话框，取消操作或执行删除操作。

## 代码演示

### 基础按钮

按钮支持 `文字按钮`、`描边按钮`、`填充按钮`、`浅亮按钮` 四种展示形态，默认是描边按钮。

`文字按钮`: 直接使用文字作为按钮。是视觉吸引力最弱的一个按钮，通常出现在表格操作栏、标题和字段旁等。

`填充按钮`: 一般用于主按钮，是用户在整个页面需要关注优先级最高的操作，引导用户关注并操作。

`描边按钮`: 常见白底加文字的形式，在视觉强调程度上弱于填充按钮，通常与填充按钮搭配成组使用

`浅亮按钮`: 浅亮状态的文字为按钮背景颜色，背景则为对应的浅色。

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="8px" flexWrap="wrap">
      <Button color="primary" variant="text">
        文字按钮
      </Button>
      <Button color="primary">描边按钮</Button>
      <Button color="primary" variant="contained">
        填充按钮
      </Button>
      <Button color="primary" variant="light">
        浅亮按钮
      </Button>
    </Stack>
  );
};
```

### 按钮大小

按钮有大、中、小三种尺寸。

通过设置 `size` 为 large small full 分别把按钮设为大、小、通栏尺寸。若不设置 `size`，则尺寸为中等尺寸。

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack spacing="8px">
      <Stack direction="row" spacing="8px">
        <Button size="small">小号按钮</Button>
        <Button size="medium">中号按钮</Button>
        <Button size="large">大号按钮</Button>
      </Stack>
      <Button size="full">通栏按钮</Button>
    </Stack>
  );
};
```

### 按钮色彩

按钮支持 `primary`、`success`、`info`、`warning`、`danger`、 `vip` 主题定义。若不设置color，则默认为`default`

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="8px" flexWrap="wrap">
      <Button variant="contained">取消</Button>
      <Button color="primary" variant="contained">
        购票
      </Button>
      <Button color="success" variant="contained">
        影城
      </Button>
      <Button color="info" variant="contained">
        电影
      </Button>
      <Button color="warning" variant="contained">
        想看
      </Button>
      <Button color="danger" variant="contained">
        特惠
      </Button>
      <Button color="vip" variant="contained">
        会员
      </Button>
    </Stack>
  );
};
```

### 禁用状态

通过 `disabled` 属性来禁用按钮，禁用状态下按钮不可点击。

```tsx
import { Button, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="8px">
      <Button variant="text" disabled>
        文字禁用
      </Button>
      <Button disabled>描边禁用</Button>
      <Button variant="contained" disabled>
        填充禁用
      </Button>
    </Stack>
  );
};
```

### 按钮带图标

因为相比纯文本来说用户对图标更敏感，所以有些时候你可能希望为某些按钮设置图标，以增强应用程序的用户体验。 例如，如果您有日期按钮，则可以使用日期图标对其进行标记。`startIcon` 位于子元素的前面，`endIcon` 位于子元素的后面。

```tsx
import React from 'react';
import { Button, Stack } from '@bifrostui/react';
import { CalendarOutlinedIcon, HeartFilledIcon } from '@bifrostui/icons';

export default () => {
  return (
    <Stack direction="row" spacing="8px">
      <Button
        startIcon={
          <HeartFilledIcon htmlColor="linear-gradient(150deg, #FFFDFC 0%, #D5B198 100%)" />
        }
      >
        Heart
      </Button>
      <Button endIcon={<CalendarOutlinedIcon />}>Date</Button>
    </Stack>
  );
};
```

## API

| 属性      | 说明                 | 类型                                                                            | 默认值    |
| --------- | -------------------- | ------------------------------------------------------------------------------- | --------- |
| variant   | 按钮类型             | `outlined` \| `text` \| `contained` \| light                                    | `text`    |
| color     | 按钮颜色             | `primary` \| `info` \| `success` \| `warning` \| `danger` \| `vip` \| `default` | `default` |
| size      | 按钮大小             | `small` \| `medium` \| `large` \| `full`                                        | `medium`  |
| disabled  | 是否禁用             | boolean                                                                         | false     |
| startIcon | 在子元素前放在的内容 | React.ReactNode                                                                 | -         |
| endIcon   | 在子元素后放在的内容 | React.ReactNode                                                                 | -         |
| openType  | 微信开放能力         | string                                                                          | -         |
| onClick   | 点击按钮时的回调     | (event: MouseEvent) => void                                                     | -         |

其他属性见 [button](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attributes)

## 样式变量

| 属性            | 说明     | 默认值                     | 全局变量                      |
| --------------- | -------- | -------------------------- | ----------------------------- |
| --border-radius | 圆角大小 | 100px                      | --bui-button-border-radius    |
| --bg-color      | 背景颜色 | --bui-color-neutral-5      | --bui-button-default-bg-color |
| --text-color    | 文字颜色 | --bui-color-fg-muted       | --bui-button-text-color       |
| --border-color  | 边框颜色 | --bui-color-border-default | --bui-button-border-color     |
| --font-weight   | 字重     | --bui-font-weight-medium   | --bui-button-font-weight      |
