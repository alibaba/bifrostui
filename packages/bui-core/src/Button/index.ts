export { default as Button, default } from './Button';
export * from './Button.types';
