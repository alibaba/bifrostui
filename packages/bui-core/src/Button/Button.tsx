import clsx from 'clsx';
import React from 'react';
import './Button.less';
import { ButtonProps } from './Button.types';

const prefixCls = 'bui-btn';

const ButtonIcon: React.FC<{
  children: React.ReactNode;
  iconPosition: 'end' | 'start';
}> = ({ children, iconPosition }) => {
  return (
    <span className={`${prefixCls}-icon-${iconPosition}`}>{children}</span>
  );
};

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      className,
      children,
      startIcon,
      endIcon,
      size,
      variant,
      color,
      disabled,
      ...others
    } = props;
    return (
      <button
        type="button"
        className={clsx(
          prefixCls,
          `${prefixCls}-${variant}`,
          `${prefixCls}-${size}`,
          {
            [`${prefixCls}-${color}`]: color,
            [`${prefixCls}-disabled`]: disabled,
          },
          className,
        )}
        ref={ref}
        {...others}
      >
        {startIcon && <ButtonIcon iconPosition="start">{startIcon}</ButtonIcon>}
        {children}
        {endIcon && <ButtonIcon iconPosition="end">{endIcon}</ButtonIcon>}
      </button>
    );
  },
);

Button.displayName = 'BuiButton';
Button.defaultProps = {
  size: 'medium',
  variant: 'outlined',
  color: 'default',
  disabled: false,
};

export default Button;
