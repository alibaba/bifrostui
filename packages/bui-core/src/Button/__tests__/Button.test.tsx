import React from 'react';
import { isConformant, render, screen, userEvent } from 'testing';
import Button from '../index';

describe('Button', () => {
  const rootClass = 'bui-btn';

  isConformant({
    Component: Button,
    displayName: 'BuiButton',
    className: rootClass,
  });
  it('renders correctly', () => {
    const { container } = render(<Button>button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('default render is medium outlined', () => {
    render(<Button>button</Button>);
    expect(screen.getByRole('button')).toHaveClass(
      `${rootClass}-outlined ${rootClass}-medium`,
    );
  });

  it('can trigger a function by being clicked', () => {
    const onClick = jest.fn();
    render(<Button onClick={onClick}>This is a button</Button>);
    userEvent.click(screen.getByRole('button'));
    expect(onClick).toHaveBeenCalled();
  });
  it('does not trigger a function when render a disabled button', () => {
    const onClick = jest.fn();
    render(<Button disabled>disabled button</Button>);
    const disabledBtn = screen.getByRole('button');
    expect(disabledBtn).toHaveClass(`${rootClass}-disabled`);
    userEvent.click(disabledBtn);
    expect(onClick).not.toHaveBeenCalled();
  });
  it.each([
    'primary',
    'warning',
    'success',
    'info',
    'danger',
    'vip',
    'default',
  ] as const)('should render correctly with color props', (color) => {
    render(<Button color={color}>{color} button</Button>);
    expect(screen.getByRole('button')).toHaveClass(`${rootClass}-${color}`);
  });
  it.each(['full', 'large', 'medium', 'medium'] as const)(
    'should render correctly with size props',
    (size) => {
      render(<Button size={size}>{size} button</Button>);
      expect(screen.getByRole('button')).toHaveClass(`${rootClass}-${size}`);
    },
  );

  it('should render a button with startIcon', () => {
    render(<Button startIcon={<span>icon</span>}>button</Button>);
    expect(screen.queryByText('icon')).toBeInTheDocument();
  });
  it('should render a button with endIcon', () => {
    render(<Button endIcon={<span>endIcon</span>}>button</Button>);
    expect(screen.queryByText('endIcon')).toBeInTheDocument();
  });
});
