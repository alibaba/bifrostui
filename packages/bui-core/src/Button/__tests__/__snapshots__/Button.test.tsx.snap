// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  class="bui-btn bui-btn-outlined bui-btn-medium bui-btn-default"
  type="button"
>
  button
</button>
`;
